src/main/java/picard/sam/markduplicates/UmiUtil.java

```diff
@@ -103,13 +103,20 @@ static boolean isTopStrand(final SAMRecord rec) {
 
         // If the read pair are aligned to different contigs we use
         // the reference index to determine relative 5' coordinate ordering.
-        if (!rec.getReferenceIndex().equals(rec.getMateReferenceIndex())) {
+        // Both the read and its mate should not have their unmapped flag set to true.
+        if (!rec.getReferenceIndex().equals(rec.getMateReferenceIndex()) && !rec.getReadUnmappedFlag() && !rec.getMateUnmappedFlag()) {
             return rec.getFirstOfPairFlag() == (rec.getReferenceIndex() < rec.getMateReferenceIndex());
         }
 
-        final int read5PrimeStart = (rec.getReadNegativeStrandFlag()) ? rec.getUnclippedEnd() : rec.getUnclippedStart();
-        final int mate5PrimeStart = (rec.getMateNegativeStrandFlag()) ? SAMUtils.getMateUnclippedEnd(rec) : SAMUtils.getMateUnclippedStart(rec);
-        return rec.getFirstOfPairFlag() == (read5PrimeStart < mate5PrimeStart);
+        final int readUnclippedStart = rec.getReadUnmappedFlag() ? 0 : rec.getUnclippedStart();
```

Comment:
I'm really nit-picking now...but reading the code for `getUnclippedStart` it seems that in extreme cases one could get zero or even negative values returned. I think that in this case the use of an Integer and `null`  are appropriate.

Reply:
Also, it isn't clear why yo are doing all these gymnastics. perhaps a comment with some diagrams of read pairs would help understand.